File: Pi-song.cpp
#include <string>
#include <fstream>
#include <iostream>
//Poco
#include <Poco\DateTime.h>
#include <Poco\NumberFormatter.h>
#include <Poco\Format.h>
using namespace std;



class pop_8_num {
public:
	//The generator of this class, aiming to generate the whole needed pi once
	pop_8_num() {
		fstream f("./source/Pi.txt");
		string buf;
		while (getline(f, buf)) {
			Pi += buf;
		}
		length = Pi.length();
	}

	auto begin() {
		return *this;
	}
	auto end() {
		return pop_8_num(length);
	}

	string operator *() {
		string result = "";
		for (int i = 0; i < 8 && current_pos != length; i++) {
			result += Pi[current_pos++];
			if (i == 3) {
				result += " ";
			}
		}
		return result;
	}
	
	auto operator ++() {
		//Do nothing
		return *this;
	}

	bool operator !=(pop_8_num other) {
		if (this->current_pos != other.current_pos)
			return true;
		else
			return false;
	}

private:
	string Pi;
	int length = 0; // Pi
	int current_pos = 0;

	pop_8_num(int i) :pop_8_num() {
		current_pos = i;
	}
};

void create_the_lrc() {
	Poco::Timespan last_time;
	Poco::Timespan span_time(3,207500);
	fstream f("Ô²ÖÜÂÊÖ®¸è.lrc", ios::trunc | ios::out);
	for (auto i : pop_8_num()) {
		last_time += span_time;
		auto hour = last_time.hours();
		auto minute = last_time.minutes();
		auto seconds = last_time.seconds();
		auto milliseconds = last_time.milliseconds();
		if (hour != 0) {
			minute += hour * 60;
		}
		if (minute >= 68 && seconds > 26) break;
		string s = Poco::format("[%02d:%02d.%02d]", minute, seconds, milliseconds/10);
		f << s << i << endl;
	}
	f.close();

}




int main() {
	create_the_lrc();
	//If you are a Linux/Unix user, just delete this.
	system("pause");

	return 0;
}

File: Pi-song.py
import datetime
import re

# Create a generator to yield 8 number of pi with a space in the center once
def pop_8_pi(precision = 10000):
    # Get pi with the precision of precision
    w = precision + 10
    b = 10 ** w
    x1 = b*4//5
    x2 = b // -239
    he = x1 + x2
    n = precision
    n *= 2
    for i in range(3, n, 2):
        x1 //= -25
        x2 //= -57121
        x = (x1+x2)//i
        he += x
    pi = he*4
    pi //= 10**10

    pi_s = str(pi)
    pi_s = pi_s[0]+"."+pi_s[1:]
    for i in range(0,len(pi_s),8):
        result = pi_s[i:i+8]
        result = result[:4]+" "+result[4:]
        yield result


def create_the_lrc():
    f = open("圆周率之歌.lrc", 'w')
    
    # Time part
    t = datetime.timedelta(0, 0)
    timediff = 3.2075
    for num_part in pop_8_pi(20000):
        t = t + datetime.timedelta(0, timediff)
        t_s = str(t)
        # Change the precision of the millisecond
        if "." in t_s:
            t_s = t_s[:-4]
        else:
            t_s += ".00"

        # Convert hour to minutes
        m = re.findall(r"\d{1,2}",t_s)
        minute = m[1]
        if m[0] == "0":
            pass
        else:
            minute = str(int(m[1])+int(m[0])*60)
        if int(minute) >= 68 and int(m[2]) > 25:
            break
        time_part = "["+minute+":"+m[2]+"."+m[3]+"]"
        f.write(time_part+num_part+"\n")
    f.close()
    print("Done")

if __name__ == "__main__":
    create_the_lrc()